File: backend/core/intel_calendar.py
"""
core/intel_calendar.py
Signal·Content·Issue·PriceMove → 통합 캘린더 이벤트 집계
"""
from __future__ import annotations

from collections import defaultdict
from datetime import date, datetime, timedelta, timezone
from typing import Any, Optional

from sqlalchemy.orm import Session

from config.database import (
    IntelContent,
    MacroSignal,
    PriceMoveCause,
    SectorSignal,
    Stock,
    StockIssue,
    StockSignal,
    WatchlistItem,
)
from core.content_scope import is_market_scope


def _empty_sentiment() -> dict[str, int]:
    return {"POSITIVE": 0, "NEUTRAL": 0, "NEGATIVE": 0}


def _bump_sentiment(counts: dict[str, int], sentiment: Optional[str]) -> None:
    key = (sentiment or "NEUTRAL").upper()
    if key not in counts:
        key = "NEUTRAL"
    counts[key] = counts.get(key, 0) + 1


def _portfolio_symbols(db: Session) -> set[str]:
    rows = db.query(Stock.symbol).filter(Stock.is_active == True, Stock.qty > 0).all()
    return {r[0] for r in rows if r[0]}


def _watchlist_symbols(db: Session) -> set[str]:
    rows = db.query(WatchlistItem.symbol).filter(WatchlistItem.symbol.isnot(None)).all()
    return {r[0] for r in rows if r[0]}


def _day_key(d: date) -> str:
    return d.strftime("%Y-%m-%d")


def _iter_dates(from_d: date, to_d: date) -> list[str]:
    out: list[str] = []
    cur = from_d
    while cur <= to_d:
        out.append(_day_key(cur))
        cur += timedelta(days=1)
    return out


def _parse_iso(s: str) -> date:
    return datetime.strptime(s[:10], "%Y-%m-%d").date()


def _content_date(c: IntelContent, date_mode: str) -> Optional[str]:
    if date_mode == "analyzed":
        dt = c.analyzed_at or c.created_at
        return dt.strftime("%Y-%m-%d") if dt else None
    if c.published_at:
        return c.published_at.strftime("%Y-%m-%d")
    if c.analyzed_at:
        return c.analyzed_at.strftime("%Y-%m-%d")
    return c.created_at.strftime("%Y-%m-%d") if c.created_at else None


def _event_dict(
    *,
    eid: str,
    day: str,
    kind: str,
    title: str,
    summary: str = "",
    sentiment: Optional[str] = None,
    sector: Optional[str] = None,
    symbol: Optional[str] = None,
    stock_name: Optional[str] = None,
    source_type: Optional[str] = None,
    content_id: Optional[int] = None,
    source_url: Optional[str] = None,
    is_portfolio: bool = False,
    is_watchlist: bool = False,
    change_pct: Optional[float] = None,
) -> dict[str, Any]:
    return {
        "id": eid,
        "date": day,
        "kind": kind,
        "title": title,
        "summary": (summary or "")[:500],
        "sentiment": (sentiment or "NEUTRAL").upper() if sentiment else None,
        "sector": sector,
        "symbol": symbol,
        "stock_name": stock_name,
        "source_type": source_type,
        "content_id": content_id,
        "source_url": source_url,
        "is_portfolio": is_portfolio,
        "is_watchlist": is_watchlist,
        "change_pct": change_pct,
    }


def _passes_filters(
    ev: dict[str, Any],
    *,
    kinds: Optional[set[str]],
    portfolio_only: bool,
    watchlist_only: bool,
) -> bool:
    if kinds and ev["kind"] not in kinds:
        return False
    if portfolio_only and not ev.get("is_portfolio"):
        return False
    if watchlist_only and not ev.get("is_watchlist"):
        return False
    return True


def collect_calendar_events(
    db: Session,
    from_date: str,
    to_date: str,
    *,
    date_mode: str = "event",
    kinds: Optional[set[str]] = None,
    portfolio_only: bool = False,
    watchlist_only: bool = False,
) -> list[dict[str, Any]]:
    """기간 내 모든 캘린더 이벤트 (flat list)."""
    from_d = _parse_iso(from_date)
    to_d = _parse_iso(to_date)
    portfolio_syms = _portfolio_symbols(db)
    watch_syms = _watchlist_symbols(db)
    stock_by_id = {s.id: s for s in db.query(Stock).all()}

    events: list[dict[str, Any]] = []

    def in_range(day: Optional[str]) -> bool:
        if not day:
            return False
        try:
            d = _parse_iso(day)
            return from_d <= d <= to_d
        except ValueError:
            return False

    if date_mode == "analyzed":
        contents = (
            db.query(IntelContent)
            .filter(IntelContent.analyzed_at.isnot(None))
            .all()
        )
        for c in contents:
            if not is_market_scope(getattr(c, "content_scope", None)):
                continue
            day = _content_date(c, "analyzed")
            if not in_range(day):
                continue
            events.append(
                _event_dict(
                    eid=f"content-{c.id}",
                    day=day,
                    kind="content",
                    title=c.source_title or f"{c.source_type} 분석",
                    summary=c.summary or "",
                    sentiment=c.sentiment,
                    source_type=c.source_type,
                    content_id=c.id,
                    source_url=c.source_url,
                )
            )
    else:
        macros = db.query(MacroSignal).filter(MacroSignal.event_date.isnot(None)).all()
        for m in macros:
            if not in_range(m.event_date):
                continue
            events.append(
                _event_dict(
                    eid=f"macro-{m.id}",
                    day=m.event_date,
                    kind="macro",
                    title=m.topic,
                    summary=m.summary or m.impact or "",
                    sentiment=m.sentiment,
                    content_id=m.content_id,
                )
            )

        sectors = db.query(SectorSignal).filter(SectorSignal.event_date.isnot(None)).all()
        for s in sectors:
            if not in_range(s.event_date):
                continue
            events.append(
                _event_dict(
                    eid=f"sector-{s.id}",
                    day=s.event_date,
                    kind="sector",
                    title=s.sector,
                    summary=s.summary or s.outlook or "",
                    sentiment=s.sentiment,
                    sector=s.sector,
                    content_id=s.content_id,
                )
            )

        stock_sigs = db.query(StockSignal).filter(StockSignal.event_date.isnot(None)).all()
        for ss in stock_sigs:
            if not in_range(ss.event_date):
                continue
            sym = (ss.symbol or "").strip()
            events.append(
                _event_dict(
                    eid=f"stock-{ss.id}",
                    day=ss.event_date,
                    kind="stock",
                    title=ss.stock_name,
                    summary=ss.summary or "",
                    sentiment=ss.sentiment,
                    symbol=sym or None,
                    stock_name=ss.stock_name,
                    content_id=ss.content_id,
                    is_portfolio=bool(ss.is_portfolio),
                    is_watchlist=sym in watch_syms if sym else False,
                )
            )

        issues = db.query(StockIssue).filter(StockIssue.event_date.isnot(None)).all()
        for iss in issues:
            if not in_range(iss.event_date):
                continue
            st = stock_by_id.get(iss.stock_id)
            sym = st.symbol if st else None
            events.append(
                _event_dict(
                    eid=f"issue-{iss.id}",
                    day=iss.event_date,
                    kind="issue",
                    title=st.name if st else "종목 이슈",
                    summary=iss.issue_summary or "",
                    sentiment=iss.sentiment,
                    symbol=sym,
                    stock_name=st.name if st else None,
                    content_id=iss.content_id,
                    is_portfolio=sym in portfolio_syms if sym else False,
                    is_watchlist=sym in watch_syms if sym else False,
                )
            )

        moves = db.query(PriceMoveCause).all()
        for mv in moves:
            if not in_range(mv.event_date):
                continue
            st = stock_by_id.get(mv.stock_id)
            sym = st.symbol if st else None
            events.append(
                _event_dict(
                    eid=f"move-{mv.id}",
                    day=mv.event_date,
                    kind="price_move",
                    title=f"{st.name if st else sym} 급변",
                    summary=mv.reason[:500] if mv.reason else "",
                    sentiment=mv.sentiment,
                    symbol=sym,
                    stock_name=st.name if st else None,
                    is_portfolio=sym in portfolio_syms if sym else False,
                    is_watchlist=sym in watch_syms if sym else False,
                    change_pct=mv.change_pct,
                )
            )

        contents = db.query(IntelContent).all()
        for c in contents:
            if not is_market_scope(getattr(c, "content_scope", None)):
                continue
            day = _content_date(c, "event")
            if not in_range(day):
                continue
            if any(e["content_id"] == c.id and e["kind"] == "content" for e in events):
                continue
            events.append(
                _event_dict(
                    eid=f"content-{c.id}",
                    day=day,
                    kind="content",
                    title=c.source_title or f"{c.source_type} 분석",
                    summary=c.summary or "",
                    sentiment=c.sentiment,
                    source_type=c.source_type,
                    content_id=c.id,
                    source_url=c.source_url,
                )
            )

    if date_mode != "analyzed":
        from config.database import EconomicCalendarEvent

        econ_rows = (
            db.query(EconomicCalendarEvent)
            .filter(
                EconomicCalendarEvent.event_date >= from_date,
                EconomicCalendarEvent.event_date <= to_date,
            )
            .all()
        )
        imp_sent = {"high": "NEGATIVE", "medium": "NEUTRAL", "low": "NEUTRAL"}
        for row in econ_rows:
            if not in_range(row.event_date):
                continue
            events.append(
                _event_dict(
                    eid=f"economic-{row.id}",
                    day=row.event_date,
                    kind="economic",
                    title=row.title,
                    summary=row.summary or "",
                    sentiment=imp_sent.get((row.importance or "").lower(), "NEUTRAL"),
                    sector=row.region,
                    source_url=row.source_url,
                )
            )

    if kinds or portfolio_only or watchlist_only:
        events = [
            e
            for e in events
            if _passes_filters(
                e,
                kinds=kinds,
                portfolio_only=portfolio_only,
                watchlist_only=watchlist_only,
            )
        ]

    events.sort(key=lambda e: (e["date"], e["kind"], e["id"]))
    return events


def _day_meta(
    events: list[dict[str, Any]],
    *,
    has_digest: bool = False,
) -> dict[str, Any]:
    sentiment = _empty_sentiment()
    by_kind: dict[str, int] = defaultdict(int)
    for e in events:
        _bump_sentiment(sentiment, e.get("sentiment"))
        by_kind[e["kind"]] += 1
    return {
        "event_count": len(events),
        "sentiment": sentiment,
        "by_kind": dict(by_kind),
        "has_digest": has_digest,
    }


def build_calendar_response(
    db: Session,
    from_date: str,
    to_date: str,
    *,
    date_mode: str = "event",
    kinds: Optional[set[str]] = None,
    portfolio_only: bool = False,
    watchlist_only: bool = False,
    include_events: bool = False,
) -> dict[str, Any]:
    events = collect_calendar_events(
        db,
        from_date,
        to_date,
        date_mode=date_mode,
        kinds=kinds,
        portfolio_only=portfolio_only,
        watchlist_only=watchlist_only,
    )
    by_day: dict[str, list[dict[str, Any]]] = defaultdict(list)
    for e in events:
        by_day[e["date"]].append(e)

    from core.intel_digest import digests_in_range

    digest_map = digests_in_range(db, from_date, to_date)

    from_d = _parse_iso(from_date)
    to_d = _parse_iso(to_date)
    days: dict[str, Any] = {}
    for day_str in _iter_dates(from_d, to_d):
        day_events = by_day.get(day_str, [])
        has_d = (
            day_str in digest_map
            and digest_map[day_str].get("status") == "ready"
        )
        meta = _day_meta(day_events, has_digest=has_d)
        if include_events:
            meta["events"] = day_events
        else:
            meta["events"] = []
        days[day_str] = meta

    kpi = _build_kpi(events, from_date, to_date)

    return {
        "from": from_date,
        "to": to_date,
        "date_mode": date_mode,
        "days": days,
        "digests": digest_map,
        "kpi": kpi,
        "total_events": len(events),
    }


def _build_kpi(events: list[dict], from_date: str, to_date: str) -> dict[str, Any]:
    sentiment = _empty_sentiment()
    sector_counts: dict[str, int] = defaultdict(int)
    content_count = 0
    portfolio_mentions = 0

    for e in events:
        _bump_sentiment(sentiment, e.get("sentiment"))
        if e["kind"] == "content":
            content_count += 1
        if e["kind"] == "sector" and e.get("sector"):
            sector_counts[e["sector"]] += 1
        if e.get("is_portfolio") or e.get("is_watchlist"):
            portfolio_mentions += 1

    total = len(events)
    top_sectors = sorted(
        [{"sector": k, "count": v} for k, v in sector_counts.items()],
        key=lambda x: -x["count"],
    )[:3]

    return {
        "from": from_date,
        "to": to_date,
        "content_count": content_count,
        "signal_count": sum(1 for e in events if e["kind"] in ("macro", "sector", "stock")),
        "total_events": total,
        "sentiment": sentiment,
        "top_sectors": top_sectors,
        "portfolio_related_count": portfolio_mentions,
    }


def build_calendar_day_response(
    db: Session,
    day: str,
    *,
    date_mode: str = "event",
    kinds: Optional[set[str]] = None,
    portfolio_only: bool = False,
    watchlist_only: bool = False,
) -> dict[str, Any]:
    events = collect_calendar_events(
        db,
        day,
        day,
        date_mode=date_mode,
        kinds=kinds,
        portfolio_only=portfolio_only,
        watchlist_only=watchlist_only,
    )
    from core.intel_digest import get_digest, serialize_digest

    digest_row = get_digest(db, day)
    has_d = bool(digest_row and digest_row.status == "ready" and digest_row.body_markdown)
    meta = _day_meta(events, has_digest=has_d)
    meta["events"] = events

    briefing = _day_briefing_from_events(day, events)
    digest_payload = None
    if digest_row and digest_row.status == "ready":
        digest_payload = serialize_digest(digest_row)
    elif digest_row and digest_row.status == "failed":
        digest_payload = serialize_digest(digest_row)

    return {
        "date": day,
        "date_mode": date_mode,
        "day": meta,
        "briefing": briefing,
        "digest": digest_payload,
        "disclaimer": "참고용 정보이며 투자 권유가 아닙니다.",
    }


def _day_briefing_from_events(day: str, events: list[dict]) -> dict[str, Any]:
    """일 뷰용 당일 요약 (기존 DailyBriefing 카드 대체)."""
    sentiment = _empty_sentiment()
    for e in events:
        _bump_sentiment(sentiment, e.get("sentiment"))
    return {
        "date": day,
        "content_count": sum(1 for e in events if e["kind"] == "content"),
        "macro_count": sum(1 for e in events if e["kind"] == "macro"),
        "sector_count": sum(1 for e in events if e["kind"] == "sector"),
        "stock_count": sum(1 for e in events if e["kind"] == "stock"),
        "issue_count": sum(1 for e in events if e["kind"] == "issue"),
        "price_move_count": sum(1 for e in events if e["kind"] == "price_move"),
        "economic_count": sum(1 for e in events if e["kind"] == "economic"),
        "sentiment_counts": sentiment,
        "event_count": len(events),
    }

File: frontend/lib/intelCalendar.ts
/**
 * 인텔리전스 캘린더 API 타입·날짜 유틸
 */
const BASE = process.env.NEXT_PUBLIC_API_URL ?? "http://localhost:8000/api";

async function fetchApi<T>(path: string, init?: RequestInit): Promise<T> {
  const res = await fetch(`${BASE}${path}`, {
    headers: { "Content-Type": "application/json" },
    ...init,
  });
  if (!res.ok) {
    const err = await res.text();
    let message = err || `HTTP ${res.status}`;
    try {
      const parsed = JSON.parse(err) as { detail?: unknown };
      if (typeof parsed.detail === "string") message = parsed.detail;
    } catch {
      /* keep */
    }
    throw new Error(message);
  }
  return res.json();
}

export type IntelCalendarKind =
  | "macro"
  | "sector"
  | "stock"
  | "content"
  | "issue"
  | "price_move"
  | "economic";

export type IntelCalendarDateMode = "event" | "analyzed";

export interface IntelCalendarEvent {
  id: string;
  date: string;
  kind: IntelCalendarKind;
  title: string;
  summary: string;
  sentiment: "POSITIVE" | "NEUTRAL" | "NEGATIVE" | null;
  sector?: string | null;
  symbol?: string | null;
  stock_name?: string | null;
  source_type?: string | null;
  content_id?: number | null;
  source_url?: string | null;
  is_portfolio?: boolean;
  is_watchlist?: boolean;
  change_pct?: number | null;
}

export interface IntelCalendarDayMeta {
  event_count: number;
  sentiment: Record<string, number>;
  by_kind: Record<string, number>;
  has_digest: boolean;
  events: IntelCalendarEvent[];
}

export interface IntelCalendarKpi {
  from: string;
  to: string;
  content_count: number;
  signal_count: number;
  total_events: number;
  sentiment: Record<string, number>;
  top_sectors: { sector: string; count: number }[];
  portfolio_related_count: number;
}

export interface IntelCalendarResponse {
  from: string;
  to: string;
  date_mode: IntelCalendarDateMode;
  days: Record<string, IntelCalendarDayMeta>;
  digests: Record<string, { title?: string; status?: string }>;
  kpi: IntelCalendarKpi;
  total_events: number;
}

export interface IntelDigest {
  date: string;
  title: string | null;
  body_markdown: string | null;
  stats: Record<string, unknown>;
  source_content_ids?: number[];
  source_signal_ids?: Record<string, number[]>;
  portfolio_highlight?: unknown;
  generated_at?: string | null;
  model?: string | null;
  status: "pending" | "ready" | "failed" | string;
  error_message?: string | null;
}

export interface IntelCalendarDayResponse {
  date: string;
  date_mode: IntelCalendarDateMode;
  day: IntelCalendarDayMeta;
  briefing: {
    date: string;
    content_count: number;
    macro_count: number;
    sector_count: number;
    stock_count: number;
    issue_count: number;
    price_move_count: number;
    economic_count?: number;
    sentiment_counts: Record<string, number>;
    event_count: number;
  };
  digest: IntelDigest | null;
  disclaimer: string;
}

export const KIND_LABELS: Record<IntelCalendarKind, string> = {
  macro: "매크로",
  sector: "섹터",
  stock: "종목",
  content: "분석",
  issue: "이슈",
  price_move: "급변",
  economic: "경제일정",
};

export const KIND_ICONS: Record<IntelCalendarKind, string> = {
  macro: "🌍",
  sector: "📊",
  stock: "📈",
  content: "🎬",
  issue: "📌",
  price_move: "⚡",
  economic: "📅",
};

function pad(n: number) {
  return String(n).padStart(2, "0");
}

export function toIsoDate(d: Date): string {
  return `${d.getFullYear()}-${pad(d.getMonth() + 1)}-${pad(d.getDate())}`;
}

export function parseIsoDate(s: string): Date {
  const [y, m, d] = s.split("-").map(Number);
  return new Date(y, m - 1, d);
}

/** 경제 일정 검색용 — 표시 기간 ±패딩 */
export function padCalendarRange(from: string, to: string): { from: string; to: string } {
  const f = parseIsoDate(from);
  f.setDate(f.getDate() - 7);
  const t = parseIsoDate(to);
  t.setDate(t.getDate() + 14);
  return { from: toIsoDate(f), to: toIsoDate(t) };
}

export function monthRange(focus: Date): { from: string; to: string } {
  const y = focus.getFullYear();
  const m = focus.getMonth();
  const from = new Date(y, m, 1);
  const to = new Date(y, m + 1, 0);
  return { from: toIsoDate(from), to: toIsoDate(to) };
}

export function kpiRange(days: number): { from: string; to: string } {
  const to = new Date();
  const from = new Date();
  from.setDate(from.getDate() - (days - 1));
  return { from: toIsoDate(from), to: toIsoDate(to) };
}

/** 월요일 시작 주간 7일 */
export function weekDates(focus: Date): string[] {
  const d = new Date(focus);
  const day = d.getDay();
  const diff = day === 0 ? -6 : 1 - day;
  d.setDate(d.getDate() + diff);
  const out: string[] = [];
  for (let i = 0; i < 7; i++) {
    out.push(toIsoDate(d));
    d.setDate(d.getDate() + 1);
  }
  return out;
}

export function monthGridCells(focus: Date): (string | null)[] {
  const y = focus.getFullYear();
  const m = focus.getMonth();
  const first = new Date(y, m, 1);
  const last = new Date(y, m + 1, 0);
  const startPad = first.getDay();
  const cells: (string | null)[] = [];
  for (let i = 0; i < startPad; i++) cells.push(null);
  for (let d = 1; d <= last.getDate(); d++) {
    cells.push(toIsoDate(new Date(y, m, d)));
  }
  while (cells.length % 7 !== 0) cells.push(null);
  return cells;
}

export interface CalendarQueryParams {
  from: string;
  to: string;
  date_mode?: IntelCalendarDateMode;
  kinds?: string;
  portfolio_only?: boolean;
  watchlist_only?: boolean;
  include_events?: boolean;
}

function buildQs(params: CalendarQueryParams): string {
  const q = new URLSearchParams();
  q.set("from", params.from);
  q.set("to", params.to);
  if (params.date_mode) q.set("date_mode", params.date_mode);
  if (params.kinds) q.set("kinds", params.kinds);
  if (params.portfolio_only) q.set("portfolio_only", "true");
  if (params.watchlist_only) q.set("watchlist_only", "true");
  if (params.include_events) q.set("include_events", "true");
  return q.toString();
}

export const intelCalendarApi = {
  getCalendar: (params: CalendarQueryParams) =>
    fetchApi<IntelCalendarResponse>(`/intel/calendar?${buildQs(params)}`),
  getDigest: (date: string) => fetchApi<IntelDigest>(`/intel/digest/${date}`),
  generateDigest: (date: string, force = false) =>
    fetchApi<{ ok: boolean; digest: IntelDigest }>("/intel/digest/generate", {
      method: "POST",
      body: JSON.stringify({ date, force }),
    }),
  backfillDigests: (from: string, to: string, force = false) =>
    fetchApi<{ ok: boolean; generated: number; results: unknown[] }>(
      "/intel/digest/backfill",
      { method: "POST", body: JSON.stringify({ from_date: from, to_date: to, force }) },
    ),
  syncEconomicCalendar: (from: string, to: string, force = false) => {
    const q = new URLSearchParams({ from, to });
    if (force) q.set("force", "true");
    return fetchApi<{ synced: boolean; count: number; message: string }>(
      `/intel/calendar/economic/sync?${q}`,
      { method: "POST" },
    );
  },
  getDay: (
    date: string,
    opts?: {
      date_mode?: IntelCalendarDateMode;
      kinds?: string;
      portfolio_only?: boolean;
      watchlist_only?: boolean;
    },
  ) => {
    const q = new URLSearchParams({ date });
    if (opts?.date_mode) q.set("date_mode", opts.date_mode);
    if (opts?.kinds) q.set("kinds", opts.kinds);
    if (opts?.portfolio_only) q.set("portfolio_only", "true");
    if (opts?.watchlist_only) q.set("watchlist_only", "true");
    return fetchApi<IntelCalendarDayResponse>(`/intel/calendar/day?${q}`);
  },
};
